### PR TITLE
Add cosine metric for Louvain clustering.

### PR DESCRIPTION
**Pull request by Hrovatin**
##### Issue
Adds cosine distance to Louvain clustering widget. 
I need this for "supplementary code" of a paper. I made the change in my own Orange version, but now it needs to be available to everyone.

##### Description of changes
This simply adds an option to the Louvain widget to use the cosine distance in sklearn's NearestNeighbors. However, NearestNeighbors uses brute force computation with cosine distance, if I understand it correctly (here is the base class https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/_base.py with specified VALID_METRIC for each nearest neighbours algorithm). Thus a modification would be required to prevent the use of cosine distance on data with too many samples. (It works fine with my data that has ~1000 samples with ~500 features.)

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by janezd** [comment]
@Hrovatin, could you experiment with different data sizes (rows and columns) and decide how large is too large?

**Comment by janezd** [comment]
@Hrovatin, I added the check for data size. Please check whether it works correctly, in particular that the error message disappears when it should, and that there is no data on output when clustering failed.

Also, set the maximum to some reasonable value. Currently it's arbitrary, 2000.

**Comment by pavlin-policar** [comment]
If you want to be really efficient you can use this little trick. The ranking of the nearest neighbors using the cosine metric is the same as the ranking using euclidean distance on row-normalized data. So, it is possible to get around using brute-force search and using the scikit-learn ball tree, but the data must be normalized first. This just gives us the ranking, and the actual distances have to be computed later on, but then you only have to calculate N\*k distances instead of N\*N. I did this in openTSNE [here](https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/nearest_neighbors.py#L90-L124) and it works fine. You can decide if it's worth the hassle though...

**Comment by janezd** [comment]
Closed in favour of #4864.
